File: chapter_api_tests/2024-04/validation/test_get_api-v1-schedules.py
import axios, { AxiosResponse } from 'axios';
import "dotenv/config";

/***************************************************************************************
 * Jest test suite for GET /api/v1/schedules
 *
 * This test suite covers:
 * 1. Input validation (query parameters)
 * 2. Response validation (status code, schema structure)
 * 3. Response headers validation
 * 4. Edge cases & limit testing
 * 5. Authentication & authorization scenarios
 ***************************************************************************************/

describe('GET /api/v1/schedules', () => {
  // Base API URL and authorization token from environment variables
  const baseURL = process.env.API_BASE_URL || '';
  const authToken = process.env.API_AUTH_TOKEN || '';

  // Create a reusable Axios instance pre-configured with base URL & headers
  const axiosInstance = axios.create({
    baseURL,
    headers: {
      Authorization: `Bearer ${authToken}`,
    },
    validateStatus: () => true, // Let us handle status code checks manually
  });

  // Helper function to check if a response has JSON content-type
  const expectJsonContentType = (headers: Record<string, string>): void => {
    expect(headers['content-type']).toContain('application/json');
  };

  // 1. Happy path: valid parameters
  it('should return 200 and a list of schedules with valid optional query parameters', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules', {
      params: {
        page: 1,
        perPage: 10,
      },
    });

    // Check status
    expect([200]).toContain(response.status);

    // Check headers
    expectJsonContentType(response.headers);

    // Basic schema checks (assuming the response should have { data: [...], meta: {...} })
    expect(response.data).toBeDefined();
    expect(response.data).toHaveProperty('data');
    expect(response.data).toHaveProperty('meta');
    expect(Array.isArray(response.data.data)).toBe(true);

    // Additional checks: page/perPage if returned, etc.
    // Example: expect(response.data.meta.page).toBe(1);
  });

  // 2. Missing optional query parameters (should still succeed)
  it('should return 200 when query parameters are omitted', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules');

    // Check status
    expect([200]).toContain(response.status);

    // Check headers
    expectJsonContentType(response.headers);

    // Basic sanity checks on response structure
    expect(response.data).toBeDefined();
    expect(response.data).toHaveProperty('data');
    expect(response.data).toHaveProperty('meta');
    expect(Array.isArray(response.data.data)).toBe(true);
  });

  // 3. Invalid query parameter type (e.g., page is string)
  it('should return 400 or 422 when "page" parameter is invalid type', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules', {
      params: {
        page: 'invalidType',
      },
    });

    // Expecting 400 or 422 for invalid parameter type
    expect([400, 422]).toContain(response.status);

    // Response headers
    expectJsonContentType(response.headers);
  });

  // 4. Out-of-bound query parameter (e.g., negative page)
  it('should return 400 or 422 when "page" parameter is negative', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules', {
      params: {
        page: -1,
      },
    });

    // Expecting 400 or 422
    expect([400, 422]).toContain(response.status);

    // Response headers
    expectJsonContentType(response.headers);
  });

  // 5. Edge case: Large perPage value
  it('should handle a large perPage value (e.g. 999999)', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules', {
      params: {
        perPage: 999999,
      },
    });

    // The API may return 200 and respond with a large dataset or an error.
    // Typically it might still return 200, so we check for both.
    // If the API absolutely disallows large perPage, it might return 400 or 422.
    expect([200, 400, 422]).toContain(response.status);

    if (response.status === 200) {
      expectJsonContentType(response.headers);
      expect(response.data).toHaveProperty('data');
      expect(response.data).toHaveProperty('meta');
      expect(Array.isArray(response.data.data)).toBe(true);
    }
  });

  // 6. Authorized vs Unauthorized requests
  it('should return 200 with valid authorization token', async () => {
    // If authToken is valid, expect 200
    // If not valid, test might fail or return 401/403
    // This is just a basic check with valid token environment setup.
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules');
    expect([200]).toContain(response.status);
    expectJsonContentType(response.headers);
  });

  it('should return 401 or 403 if no authorization token is provided', async () => {
    // Create a second instance without the Authorization header
    const noAuthAxios = axios.create({
      baseURL,
      validateStatus: () => true,
    });

    const response: AxiosResponse = await noAuthAxios.get('/api/v1/schedules');

    // Expect 401 or 403
    expect([401, 403]).toContain(response.status);
  });

  // 7. Handling empty or no results found scenario
  // This test is conceptual, as it depends on the state of the system.
  // One approach: request a page number so large that no results should exist.
  it('should return an empty data array if no schedules are found', async () => {
    const response: AxiosResponse = await axiosInstance.get('/api/v1/schedules', {
      params: {
        page: 999999, // hoping to get an empty list
      },
    });

    // Ensure valid 200 status
    expect([200]).toContain(response.status);
    expectJsonContentType(response.headers);

    // Check for empty result array if it is truly empty
    expect(response.data).toBeDefined();
    expect(response.data).toHaveProperty('data');

    if (Array.isArray(response.data.data)) {
      // If no schedules exist on this page, we expect an empty array
      // (Can't guarantee environment data, so we do a safe check)
      if (response.data.data.length === 0) {
        expect(response.data.data.length).toBe(0);
      } else {
        // If we do have data, at least it is a valid array
        expect(Array.isArray(response.data.data)).toBe(true);
      }
    }
  });

  // 8. Server error (5xx) scenario
  // Hard to force a 5xx in a test environment without special setup.
  // Below is a placeholder in case you can simulate a server error.
  // it('should handle 500 internal server error scenario', async () => {
  //   // If the server can be forced to error, test it here.
  //   // In normal situations, you might mock or intercept the request.
  //   // Or test a known edge case for the server.
  // });
});

File: chapter_api_tests/2024-04/validation/test_get_api-v1-schedules.ts
import axios, { AxiosError, AxiosResponse } from 'axios';
import { describe, it, expect, beforeAll } from '@jest/globals';

/*************************************************
 * Test Suite for GET /api/v1/schedules
 * Method: GET
 * Path: /api/v1/schedules
 * Description: List all schedules with optional pagination.
 *************************************************/

describe('GET /api/v1/schedules', () => {
  let baseURL: string;
  let token: string;

  beforeAll(() => {
    // Load environment variables
    baseURL = process.env.API_BASE_URL || 'http://localhost:3000';
    token = process.env.API_AUTH_TOKEN || '';
  });

  /**
   * Helper function to create an axios instance with common headers
   */
  const getAxiosInstance = (authToken?: string) => {
    return axios.create({
      baseURL,
      headers: {
        'Content-Type': 'application/json',
        ...(authToken ? { Authorization: `Bearer ${authToken}` } : {}),
      },
      validateStatus: () => true, // We handle status checks manually
    });
  };

  /**
   * 1. Input Validation
   *    - Test valid/invalid query params.
   */
  it('should return 200 OK with valid query parameters (page, perPage)', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules', {
      params: {
        page: 1,
        perPage: 10,
      },
    });

    // Expect a 200 status for valid query params
    expect(response.status).toBe(200);
    // Check for application/json header
    expect(response.headers['content-type']).toContain('application/json');
    // Check for a valid response body structure (partial, as example)
    expect(response.data).toBeDefined();
    // Example: If the schema includes a schedules array, verify
    // Adjust the property checks below to match the actual schema from #/components/schemas/ListSchedulesResult
    // expect(Array.isArray(response.data.schedules)).toBe(true);
  });

  it('should allow no query parameters and return 200 with a default page result', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules');

    expect(response.status).toBe(200);
    expect(response.headers['content-type']).toContain('application/json');
    expect(response.data).toBeDefined();
    // Perform partial schema checks
  });

  it('should return 400 or 422 when page parameter is invalid (e.g., a string)', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules', {
      params: {
        page: 'invalid',
      },
    });

    // The API may return 400 or 422 for invalid parameters.
    expect([400, 422]).toContain(response.status);
    expect(response.headers['content-type']).toContain('application/json');
  });

  it('should return 400 or 422 when perPage parameter is invalid (e.g., negative)', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules', {
      params: {
        perPage: -10,
      },
    });

    expect([400, 422]).toContain(response.status);
    expect(response.headers['content-type']).toContain('application/json');
  });

  /**
   * 2. Response Validation
   *    - Validate 200 success structure, error codes, etc.
   */
  it('should match the expected 200 response schema for a valid request', async () => {
    const instance = getAxiosInstance(token);
    const response: AxiosResponse = await instance.get('/api/v1/schedules');

    expect(response.status).toBe(200);
    expect(response.headers['content-type']).toContain('application/json');

    // Example partial schema validation
    // Adjust to match #/components/schemas/ListSchedulesResult
    // expect(Array.isArray(response.data.schedules)).toBe(true);
    // expect(response.data.page).toBeDefined();
    // expect(response.data.total).toBeDefined();
  });

  /**
   * 3. Response Headers Validation
   *    - Check "Content-Type", etc.
   */
  it('should include Content-Type header in the response', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules');

    expect(response.status).toBe(200);
    expect(response.headers['content-type']).toContain('application/json');
  });

  /**
   * 4. Edge Case & Limit Testing
   */
  it('should return an empty array or valid structure if no schedules exist (edge case)', async () => {
    // This test presumes a scenario where the database might be empty.
    // If your environment always has schedules, adapt as needed.
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules', {
      params: {
        page: 999999, // Large page number may return an empty list
        perPage: 100,
      },
    });

    // Expect a successful response with potential empty data.
    expect(response.status).toBe(200);
    expect(response.data).toBeDefined();
    // Adjust property checks based on your actual response schema.
    // Example:
    // expect(Array.isArray(response.data.schedules)).toBe(true);
    // expect(response.data.schedules.length).toBe(0);
  });

  it('should ensure proper handling with extremely large perPage value', async () => {
    const instance = getAxiosInstance(token);

    const response: AxiosResponse = await instance.get('/api/v1/schedules', {
      params: {
        perPage: 999999999, // A large integer to test boundaries.
      },
    });

    // The API might still return 200 or possibly 400 if it's out of range.
    // Adjust expectations based on your API's behavior.
    expect([200, 400, 422]).toContain(response.status);
  });

  it('should handle server errors gracefully if the server returns a 500 (hypothetical test)', async () => {
    // This test assumes you might force a 500 error by some specific input or environment.
    // Adjust or remove this if 500 is not easily triggered.
    const instance = getAxiosInstance(token);

    // Example only: Not guaranteed to trigger a 500.
    // In a real environment, you might have a special test setup to provoke a server error.
    try {
      const response = await instance.get('/api/v1/schedules', {
        params: {
          page: -999999, // Possibly invalid enough to cause a server error in some implementations.
        },
      });
      // If the server does not return 500, check acceptable alternate statuses.
      expect([200, 400, 422]).toContain(response.status);
    } catch (err) {
      const error = err as AxiosError;
      // Check if we indeed got a 500
      if (error.response) {
        expect(error.response.status).toBe(500);
      }
    }
  });

  /**
   * 5. Testing Authorization & Authentication
   *    - Test valid, invalid, and missing credentials.
   */
  it('should return 200 with a valid token', async () => {
    // Assuming token is valid if provided.
    if (!token) {
      console.warn('No valid API_AUTH_TOKEN found; skipping test.');
      return;
    }

    const instance = getAxiosInstance(token);
    const response: AxiosResponse = await instance.get('/api/v1/schedules');

    expect(response.status).toBe(200);
    expect(response.headers['content-type']).toContain('application/json');
  });

  it('should return 401 or 403 for missing or invalid token', async () => {
    const instance = getAxiosInstance('InvalidOrMissingToken');
    const response: AxiosResponse = await instance.get('/api/v1/schedules');

    // The API might return 401 or 403.
    expect([401, 403]).toContain(response.status);
  });
});
